### PR TITLE
Feat/batch inference operator

### PR DESCRIPTION
**Pull request by vlad-tokarev**
Kubernetes controller for #500

### CODE REVIEW

File: packages/operator/api/v1alpha1/batchinferencejob_types.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	// can be escaped with a double $$, ie: $$(VAR_NAME). Escaped references will never be expanded,
+	// regardless of whether the variable exists or not.
+	// Cannot be updated.
+	// More info: https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/#running-a-command-in-a-shell
+	Args []string `json:"args"`
+	// InputConnection is name of connection to object storage bucket where input data are expected
+	InputConnection string `json:"inputConnection"`
+	// InputPath is a source directory for BatchInferenceJob input data
+	// relative to bucket root of InputConnection
+	InputPath string `json:"inputPath"`
```
Comment (keshamin) [inline packages/operator/api/v1alpha1/batchinferencejob_types.go:65]: _Minor_
Intuitively, it seems attractive to create a small special structure like:
```
type ObjectLocation struct {
  ConnectionID string
  Path string
}
```
and use it for `Input`, `Output` and `Model`.

Comment (vlad-tokarev) [reply packages/operator/api/v1alpha1/batchinferencejob_types.go:65]: Let's leave it as is

**Review by keshamin — APPROVED** [review]

**Review by easokol — APPROVED** [review]

File: helms/odahu-flow-core/templates/operator/generated/odahuflow.odahu.org_batchinferencejobs.yaml [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+            outputConnection:
+              description: OutputConnection is name of connection to object storage
+                bucket where results should be saved
+              type: string
+            outputPath:
+              description: OutputPath is a destination directory for BatchInferenceJob
+                results relative to bucket root of OutputConnection
+              type: string
+            requestId:
+              description: BatchRequestID is unique identifier for BatchInferenceJob
```
Comment (karbyshevds) [inline helms/odahu-flow-core/templates/operator/generated/odahuflow.odahu.org_batchinferencejobs.yaml:108]: Looks like some dots are missed in this text wall.

File: helms/odahu-flow-core/templates/operator/generated/odahuflow.odahu.org_batchinferencejobs.yaml [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                that helps to correlate between Model input, model output and feedback
+                Take into account that it is not the same as kubeflow InferenceRequest
+                id Each BatchInferenceJob can process more than one InferenceRequest
+                (delivered in separate input file) So each BatchRequestID has set
+                of corresponding InferenceRequest and their IDs
+              type: string
+            resources:
+              description: Resources for model container The same format like k8s
```
Comment (karbyshevds) [inline helms/odahu-flow-core/templates/operator/generated/odahuflow.odahu.org_batchinferencejobs.yaml:116]:
```suggestion
              description: Resources for model container The same format as k8s
```

File: helms/odahu-flow-core/templates/operator/generated/odahuflow.odahu.org_batchinferencejobs.yaml [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+          properties:
+            message:
+              type: string
+            podName:
+              type: string
+            reason:
+              type: string
+            state:
+              description: 'INSERT ADDITIONAL STATUS FIELD - define observed state
+                of cluster Important: Run "make" to regenerate code after modifying
```
Comment (karbyshevds) [inline helms/odahu-flow-core/templates/operator/generated/odahuflow.odahu.org_batchinferencejobs.yaml:171]:
```suggestion
                of cluster. Important: Run "make" to regenerate code after modifying
```

File: packages/operator/controllers/batchinferencejob_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,322 @@
+//
+//    Copyright 2019 EPAM Systems
```
Comment (karbyshevds) [inline packages/operator/controllers/batchinferencejob_controller_test.go]: Can we remove date mention?

File: packages/operator/docs/docs.go [hunk 6]
```diff
@@ -2076,6 +2076,46 @@ var doc = `{
...
+                },
+                "nodePools": {
+                    "description": "Node pools to run deployments",
+                    "type": "array",
+                    "items": {
+                        "$ref": "#/definitions/NodePool"
+                    }
+                },
+                "rcloneImage": {
+                    "description": "RClone image that will be used to sync data with object storage",
```
Comment (karbyshevds) [inline packages/operator/docs/docs.go:2098]:
```suggestion
                    "description": "Rclone image that will be used to sync data with object storage",
```

File: packages/operator/docs/docs.go [hunk 7]
```diff
@@ -2076,6 +2076,46 @@ var doc = `{
...
+                "tolerations": {
+                    "description": "Kubernetes tolerations for batch jobs",
+                    "type": "string"
+                },
+                "toolsImage": {
+                    "description": "ODAHU tools image",
+                    "type": "string"
+                },
+                "toolsSecret": {
+                    "description": "ODAHU tools secret name with config",
```
Comment (karbyshevds) [inline packages/operator/docs/docs.go]:
```suggestion
                    "description": "Name of the secret that contains ODAHU tools config",
```

File: packages/operator/docs/docs.go [hunk 8]
```diff
@@ -2076,6 +2076,46 @@ var doc = `{
...
+                "enabled": {
+                    "description": "Enable batch API/operator",
+                    "type": "boolean"
+                },
+                "namespace": {
+                    "description": "Kubernetes namespace, where BatchInferenceService and BatchInferenceJob will be created",
+                    "type": "string"
+                },
+                "nodePools": {
+                    "description": "Node pools to run deployments",
```
Comment (karbyshevds) [inline packages/operator/docs/docs.go]:
```suggestion
                    "description": "Node pools to run batch jobs",
```

File: packages/operator/docs/docs.go [hunk 9]
```diff
@@ -2076,6 +2076,46 @@ var doc = `{
...
+                    "items": {
+                        "$ref": "#/definitions/NodePool"
+                    }
+                },
+                "rcloneImage": {
+                    "description": "RClone image that will be used to sync data with object storage",
+                    "type": "string"
+                },
+                "timeout": {
+                    "description": "Timeout for full training process",
```
Comment (karbyshevds) [inline packages/operator/docs/docs.go]: Is `training` timeout correct here?

File: packages/operator/docs/swagger.json [hunk 10]
```diff
@@ -2060,6 +2060,46 @@
...
+                "enabled": {
+                    "description": "Enable batch API/operator",
+                    "type": "boolean"
+                },
+                "namespace": {
+                    "description": "Kubernetes namespace, where BatchInferenceService and BatchInferenceJob will be created",
+                    "type": "string"
+                },
+                "nodePools": {
+                    "description": "Node pools to run deployments",
```
Comment (karbyshevds) [inline packages/operator/docs/swagger.json]:
```suggestion
                    "description": "Node pools to run batch jobs",
```

File: packages/operator/docs/swagger.json [hunk 11]
```diff
@@ -2060,6 +2060,46 @@
...
+                    "type": "array",
+                    "items": {
+                        "$ref": "#/definitions/NodePool"
+                    }
+                },
+                "rcloneImage": {
+                    "description": "RClone image that will be used to sync data with object storage",
```
Comment (karbyshevds) [inline packages/operator/docs/swagger.json:2082]:
```suggestion
                    "description": "Rclone image that will be used to sync data with object storage",
```

File: packages/operator/docs/swagger.json [hunk 12]
```diff
@@ -2060,6 +2060,46 @@
...
+                "tolerations": {
+                    "description": "Kubernetes tolerations for batch jobs",
+                    "type": "string"
+                },
+                "toolsImage": {
+                    "description": "ODAHU tools image",
+                    "type": "string"
+                },
+                "toolsSecret": {
+                    "description": "ODAHU tools secret name with config",
```
Comment (karbyshevds) [inline packages/operator/docs/swagger.json]:
```suggestion
                    "description": "Name of the secret that contains ODAHU tools config",
```

File: packages/operator/docs/swagger.json [hunk 13]
```diff
@@ -2060,6 +2060,46 @@
...
+                    "items": {
+                        "$ref": "#/definitions/NodePool"
+                    }
+                },
+                "rcloneImage": {
+                    "description": "RClone image that will be used to sync data with object storage",
+                    "type": "string"
+                },
+                "timeout": {
+                    "description": "Timeout for full training process",
```
Comment (karbyshevds) [inline packages/operator/docs/swagger.json]: Is `training` timeout correct here?

**Review by karbyshevds — CHANGES_REQUESTED** [review]
All `packages/operator/docs/docs.go` suggests are also covers `swagger.yaml`

File: packages/operator/controllers/batchinferencejob_controller_test.go (hunk 5 above)
Comment (vlad-tokarev) [reply packages/operator/controllers/batchinferencejob_controller_test.go]: Let's just update the date for now. In the future, we can consider removing the date from the license at all

**Review by karbyshevds — APPROVED** [review]

**Pull request merged**